## Student Performance Analysis

In [1]:
import pandas as pd
df=pd.read_csv("StudentData.csv")

### Data Inspection

In [2]:
df.head(5)

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate
0,1,Bronnie,Female,13,New York,85,81.0,62.0,70.0,87,2023-07-07
1,2,Lemmie,Male,15,San Antonio,77,41.0,77.0,50.0,98,2023-07-09
2,3,Danya,Female,14,New York,85,87.0,90.0,96.0,99,2023-07-09
3,4,Denna,Female,14,New York,60,65.0,82.0,65.0,99,2023-07-10
4,5,Jocelin,Male,14,San Antonio,43,80.0,60.0,65.0,99,2023-07-11


In [3]:
df.shape

(25, 11)

In [4]:
df.columns

Index(['ID', 'Name', 'Gender', 'Age', 'City', 'Science', 'English', 'History',
       'Maths', 'Attendance', 'AdmissionDate'],
      dtype='object')

In [5]:
df.dtypes

ID                 int64
Name              object
Gender            object
Age                int64
City              object
Science            int64
English          float64
History          float64
Maths            float64
Attendance         int64
AdmissionDate     object
dtype: object

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ID             25 non-null     int64  
 1   Name           25 non-null     object 
 2   Gender         25 non-null     object 
 3   Age            25 non-null     int64  
 4   City           25 non-null     object 
 5   Science        25 non-null     int64  
 6   English        24 non-null     float64
 7   History        24 non-null     float64
 8   Maths          24 non-null     float64
 9   Attendance     25 non-null     int64  
 10  AdmissionDate  25 non-null     object 
dtypes: float64(3), int64(4), object(4)
memory usage: 2.3+ KB


In [7]:
df.describe()

,ID,Age,Science,English,History,Maths,Attendance
count,25.000000,25.000000,25.000000,24.000000,24.000000,24.000000,25.000000
mean,12.080000,13.840000,71.160000,74.250000,75.875000,71.166667,93.960000
std,7.772387,0.746101,17.425269,17.343712,14.715601,15.854410,5.586293
min,1.000000,13.000000,40.000000,41.000000,44.000000,45.000000,83.000000
25%,5.000000,13.000000,60.000000,61.500000,65.750000,60.000000,89.000000
50%,12.000000,14.000000,76.000000,80.500000,78.500000,71.000000,95.000000
75%,19.000000,14.000000,85.000000,87.000000,88.500000,80.000000,99.000000
max,25.000000,15.000000,98.000000,96.000000,96.000000,96.000000,100.000000


### Average Marks: Science(71.16), English(74.25), History(75.87), Maths(71.16)
### Max Marks: Science(98.00), English(96.00), History(96.00), Maths(96.00)	

In [8]:
df.isnull()

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate
0,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False
5,False,False,False,False,False,False,False,False,False,False,False
6,False,False,False,False,False,False,False,True,False,False,False
7,False,False,False,False,False,False,False,False,False,False,False
8,False,False,False,False,False,False,False,False,False,False,False
9,False,False,False,False,False,False,False,False,False,False,False


In [9]:
df.isnull().sum()

ID               0
Name             0
Gender           0
Age              0
City             0
Science          0
English          1
History          1
Maths            1
Attendance       0
AdmissionDate    0
dtype: int64

### Data Cleaning

In [10]:
df=df.drop_duplicates()

In [11]:
df

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate
0,1,Bronnie,Female,13,New York,85,81.0,62.0,70.0,87,2023-07-07
1,2,Lemmie,Male,15,San Antonio,77,41.0,77.0,50.0,98,2023-07-09
2,3,Danya,Female,14,New York,85,87.0,90.0,96.0,99,2023-07-09
3,4,Denna,Female,14,New York,60,65.0,82.0,65.0,99,2023-07-10
4,5,Jocelin,Male,14,San Antonio,43,80.0,60.0,65.0,99,2023-07-11
5,6,Malissa,Female,14,Los Angeles,98,51.0,85.0,76.0,85,2023-07-12
6,7,Ichabod,Female,14,Dallas,69,74.0,NaN,60.0,100,2023-07-13
7,8,Beverlie,Male,14,Los Angeles,78,80.0,85.0,80.0,95,2023-07-14
8,9,Corrine,Male,15,Los Angeles,78,89.0,88.0,90.0,98,2023-07-15
9,10,Tate,Male,15,San Antonio,70,50.0,60.0,55.0,99,2023-07-20


In [12]:
numeric_cols = ["Science", "English", "Maths", "History", "Attendance"]

df[numeric_cols] = df[numeric_cols].fillna(
    df[numeric_cols].mean()
)

### Filled the missing marks in a subject with average marks in that subject.

In [13]:
df.isnull().sum().sum()

np.int64(0)

In [14]:
df["Name"]=df["Name"].str.title()

In [15]:
df

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate
0,1,Bronnie,Female,13,New York,85,81.000000,62.000000,70.000000,87,2023-07-07
1,2,Lemmie,Male,15,San Antonio,77,41.000000,77.000000,50.000000,98,2023-07-09
2,3,Danya,Female,14,New York,85,87.000000,90.000000,96.000000,99,2023-07-09
3,4,Denna,Female,14,New York,60,65.000000,82.000000,65.000000,99,2023-07-10
4,5,Jocelin,Male,14,San Antonio,43,80.000000,60.000000,65.000000,99,2023-07-11
5,6,Malissa,Female,14,Los Angeles,98,51.000000,85.000000,76.000000,85,2023-07-12
6,7,Ichabod,Female,14,Dallas,69,74.000000,74.590909,60.000000,100,2023-07-13
7,8,Beverlie,Male,14,Los Angeles,78,80.000000,85.000000,80.000000,95,2023-07-14
8,9,Corrine,Male,15,Los Angeles,78,89.000000,88.000000,90.000000,98,2023-07-15
9,10,Tate,Male,15,San Antonio,70,50.000000,60.000000,55.000000,99,2023-07-20


In [16]:
df["AdmissionDate"]=pd.to_datetime(df["AdmissionDate"])

In [17]:
df["Year"]=df["AdmissionDate"].dt.year

In [18]:
df["Month"]=df["AdmissionDate"].dt.month

In [19]:
df

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate,Year,Month
0,1,Bronnie,Female,13,New York,85,81.000000,62.000000,70.000000,87,2023-07-07,2023,7
1,2,Lemmie,Male,15,San Antonio,77,41.000000,77.000000,50.000000,98,2023-07-09,2023,7
2,3,Danya,Female,14,New York,85,87.000000,90.000000,96.000000,99,2023-07-09,2023,7
3,4,Denna,Female,14,New York,60,65.000000,82.000000,65.000000,99,2023-07-10,2023,7
4,5,Jocelin,Male,14,San Antonio,43,80.000000,60.000000,65.000000,99,2023-07-11,2023,7
5,6,Malissa,Female,14,Los Angeles,98,51.000000,85.000000,76.000000,85,2023-07-12,2023,7
6,7,Ichabod,Female,14,Dallas,69,74.000000,74.590909,60.000000,100,2023-07-13,2023,7
7,8,Beverlie,Male,14,Los Angeles,78,80.000000,85.000000,80.000000,95,2023-07-14,2023,7
8,9,Corrine,Male,15,Los Angeles,78,89.000000,88.000000,90.000000,98,2023-07-15,2023,7
9,10,Tate,Male,15,San Antonio,70,50.000000,60.000000,55.000000,99,2023-07-20,2023,7


### Feature Engineering

In [20]:
df=df.copy()

In [21]:
df["Total"]=(df["Science"]+df["English"]+df["History"]+df["Maths"])

In [22]:
df["Average"]=df["Total"]/4

In [23]:
df["Result"]=df["Average"].apply(lambda x:"Pass" if x>=60 else "Fail")

In [24]:
df["Grade"]=df["Average"].apply(lambda x:"A" if x>=90 else "B" if x>=80 else "C" if x>=70 else "D" if x>=60 else "F")

### Data Analysis

In [25]:
df.loc[df["Total"]==df["Total"].max()]

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate,Year,Month,Total,Average,Result,Grade
14,15,Missy,Female,13,New York,98,89.0,84.0,90.0,84,2024-02-01,2024,2,361.0,90.25,Pass,A


### Missy from New York has highest total marks i.e 361

In [26]:
df.groupby("City")["Average"].mean()

City
Dallas         69.779545
Houston        65.750000
Los Angeles    75.100000
New York       75.337121
San Antonio    67.945455
Name: Average, dtype: float64

### Students from NewYork has highest average total marks while Dallas students has min average total.

In [27]:
df.groupby("City")["Attendance"].mean()

City
Dallas         96.600000
Houston        88.500000
Los Angeles    91.200000
New York       91.833333
San Antonio    96.800000
Name: Attendance, dtype: float64

### San Antonio students has highest attendance percentage.

In [28]:
df.groupby("Result")["Name"].count()

Result
Fail     4
Pass    19
Name: Name, dtype: int64

### Total 19 students pass out of 23 students

In [29]:
df.groupby("Gender")["Maths"].mean()

Gender
Female    72.490909
Male      66.153846
Name: Maths, dtype: float64

### Female students have highest average math marks.

In [30]:
df.sort_values(by="Average",ascending=False).head(5)

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate,Year,Month,Total,Average,Result,Grade
14,15,Missy,Female,13,New York,98,89.0,84.0,90.0,84,2024-02-01,2024,2,361.0,90.25,Pass,A
2,3,Danya,Female,14,New York,85,87.0,90.0,96.0,99,2023-07-09,2023,7,358.0,89.50,Pass,B
8,9,Corrine,Male,15,Los Angeles,78,89.0,88.0,90.0,98,2023-07-15,2023,7,345.0,86.25,Pass,B
24,25,Jone,Male,14,Houston,89,87.0,95.0,70.0,88,2024-03-09,2024,3,341.0,85.25,Pass,B
12,13,Merrick,Male,15,Los Angeles,65,89.0,96.0,80.0,83,2023-07-25,2023,7,330.0,82.50,Pass,B


In [31]:
df[["Science", "English", "History", "Maths"]].mean()

Science    69.956522
English    73.090909
History    74.590909
Maths      68.909091
dtype: float64

### Average Marks: Science: 69.95 , English:73.09, History:74.59, Maths:68.91

In [32]:
df.groupby("Year")["Name"].count()

Year
2023    12
2024    11
Name: Name, dtype: int64

### In 2023 more student enrolled than 2024.

### Pivot Tables

In [33]:
df.pivot_table(values="Average",index="City")

,Average
City,
Dallas,69.779545
Houston,65.750000
Los Angeles,75.100000
New York,75.337121
San Antonio,67.945455


In [34]:
df.pivot_table(values="Science",index="Gender")

,Science
Gender,
Female,76.4
Male,65.0


### Female has highest average science marks than male.

In [35]:
df.pivot_table(values="Attendance",index=["City","Gender"])

Attendance
City        Gender            
Dallas      Female   95.333333
            Male     98.500000
Houston     Male     88.500000
Los Angeles Female   85.000000
            Male     92.750000
New York    Female   93.000000
            Male     86.000000
San Antonio Female   95.000000
            Male     97.250000

### Male attendance in most of the cities shows higher percentage than female.

### Crosstabs

In [36]:
pd.crosstab(df["City"],df["Gender"])

Gender,Female,Male
City,,
Dallas,3,2
Houston,0,2
Los Angeles,1,4
New York,5,1
San Antonio,1,4


### More students belongs to the city New York in which Females rate is higher than males.

In [37]:
pd.crosstab(df["Grade"],df["Result"])

Result,Fail,Pass
Grade,,
A,0,1
B,0,6
C,0,6
D,0,6
F,4,0


### Only 1 student got A Grade

### Sorting & Filtering

In [38]:
df.loc[df["Average"]>85]


,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate,Year,Month,Total,Average,Result,Grade
2,3,Danya,Female,14,New York,85,87.0,90.0,96.0,99,2023-07-09,2023,7,358.0,89.50,Pass,B
8,9,Corrine,Male,15,Los Angeles,78,89.0,88.0,90.0,98,2023-07-15,2023,7,345.0,86.25,Pass,B
14,15,Missy,Female,13,New York,98,89.0,84.0,90.0,84,2024-02-01,2024,2,361.0,90.25,Pass,A
24,25,Jone,Male,14,Houston,89,87.0,95.0,70.0,88,2024-03-09,2024,3,341.0,85.25,Pass,B


In [39]:
df.loc[(df["City"]=="New York") & (df["Attendance"]>90)]

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate,Year,Month,Total,Average,Result,Grade
2,3,Danya,Female,14,New York,85,87.0,90.0,96.0,99,2023-07-09,2023,7,358.0,89.5,Pass,B
3,4,Denna,Female,14,New York,60,65.0,82.0,65.0,99,2023-07-10,2023,7,272.0,68.0,Pass,D
18,19,Opalina,Female,13,New York,62,77.0,75.0,72.0,96,2024-03-03,2024,3,286.0,71.5,Pass,C


In [40]:
df.sort_values(
    by="Average",
    ascending=False
)

,ID,Name,Gender,Age,City,Science,English,History,Maths,Attendance,AdmissionDate,Year,Month,Total,Average,Result,Grade
14,15,Missy,Female,13,New York,98,89.000000,84.000000,90.000000,84,2024-02-01,2024,2,361.000000,90.250000,Pass,A
2,3,Danya,Female,14,New York,85,87.000000,90.000000,96.000000,99,2023-07-09,2023,7,358.000000,89.500000,Pass,B
8,9,Corrine,Male,15,Los Angeles,78,89.000000,88.000000,90.000000,98,2023-07-15,2023,7,345.000000,86.250000,Pass,B
24,25,Jone,Male,14,Houston,89,87.000000,95.000000,70.000000,88,2024-03-09,2024,3,341.000000,85.250000,Pass,B
12,13,Merrick,Male,15,Los Angeles,65,89.000000,96.000000,80.000000,83,2023-07-25,2023,7,330.000000,82.500000,Pass,B
7,8,Beverlie,Male,14,Los Angeles,78,80.000000,85.000000,80.000000,95,2023-07-14,2023,7,323.000000,80.750000,Pass,B
15,16,Gilberta,Male,15,San Antonio,79,90.000000,90.000000,64.000000,93,2024-02-02,2024,2,323.000000,80.750000,Pass,B
20,21,Amery,Male,13,Dallas,76,96.000000,67.000000,80.000000,97,2024-03-05,2024,3,319.000000,79.750000,Pass,C
5,6,Malissa,Female,14,Los Angeles,98,51.000000,85.000000,76.000000,85,2023-07-12,2023,7,310.000000,77.500000,Pass,C
13,14,Gracie,Female,13,San Antonio,93,76.000000,70.000000,68.909091,95,2024-02-01,2024,2,307.909091,76.977273,Pass,C


In [41]:
df.to_csv("student_performance_cleaned.csv",index=False)